packages/opensrp-reducer-factory/src/index.ts

```diff
@@ -0,0 +1,242 @@
+ *
+ * @param {string} reducerName - generic reducer name
+ * @returns {object} - the state
+ */
+export const reducerFactory = <ObjectType>(reducerName: string) =>
+  /** the objects reducer function */
+  function reducer(
+    state: ImmutableObjectState<ObjectType> = initialState,
+    action: ItemsActionTypes<ObjectType>
+  ): ImmutableObjectState<ObjectType> {
+    const actionReducerName = action.reducerName;
+    if (reducerName !== actionReducerName) {
+      return state;
+    }
+    switch (action.type) {
+      case FETCHED:
+        return SeamlessImmutable({
+          ...state,
+          objectsById: { ...state.objectsById, ...action.objectsById },
+        });
+      case REMOVE:
+        return SeamlessImmutable({
+          ...state,
+          objectsById: action.objectsById,
+        });
+      case SET_TOTAL_RECORDS:
+        return SeamlessImmutable({
+          ...state,
+          totalRecords: action.totalRecords,
+        });
+      default:
+        return state;
+    }
+  };
+
+// Selectors
+
+/** factory function that creates selector
+ *  ObjectType - generic type - object type being handled by this function
+ *
+ *  @param {string} reducerName - the reducerName
+ *  @returns {function()} - function that returns the state
+ */
+export const getItemsByIdFactory = <ObjectType>(
+  reducerName: string
+): ((state: Partial<Store>) => Dictionary<ObjectType>) => {
+  /** returns all objects in the store as values whose keys are their respective ids
+   *
+   * @param {object} state - the redux store
+   * @returns {object} - an object whose keys are the ids
+   */
+  return function (state: Partial<Store>): Dictionary<ObjectType> {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    return (state as any)[reducerName].objectsById;
+  };
+};
+
+/** factory function that creates selector
+ *
+ * @param {string} reducerName - name of the reducer
+ * @returns {function()} - an array of object type being handled by this function
+ */
+export const getItemsArrayFactory = <ObjectType>(reducerName: string) =>
+  /** gets an array of objects
+   *
+   * @param {Dictionary} state - the redux store
+   * @returns {object[]} - an array of objs
+   */
+  function (state: Partial<Store>): ObjectType[] {
+    const getItemsById = getItemsByIdFactory<ObjectType>(reducerName);
+    return values<ObjectType>(getItemsById(state));
+  };
+
+/** factory function that creates selector
+ *
+ * @param {string} reducerName -  name of reducer
+ * @returns {function()} - object type being handled by this function
+ */
+export const getItemByIdFactory = <ObjectType>(reducerName: string) =>
+  /** get a specific object by their id
+   *
+   * @param {Dictionary} state - the redux store
+   * @returns {Dictionary | null} a obj if the id is found else null
+   */
+  function (state: Partial<Store>, id: string): ObjectType | null {
+    return get(getItemsByIdFactory<ObjectType>(reducerName)(state), id) || null;
```

Comment:
Is it possible to make the identifier field configurable so that it is not hard coded to "id"?

Reply:
@moshthepitt id appears to be a variable, its not hard-coded